homeassistant/components/lcn/websocket.py

```diff
@@ -0,0 +1,419 @@
+"""LCN Websocket API."""
+
+from __future__ import annotations
+
+from copy import deepcopy
+from typing import Final
+
+import lcn_frontend as lcn_panel
+import voluptuous as vol
+
+from homeassistant.components import panel_custom, websocket_api
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_ADDRESS,
+    CONF_DEVICES,
+    CONF_DOMAIN,
+    CONF_ENTITIES,
+    CONF_ID,
+    CONF_NAME,
+    CONF_RESOURCE,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import device_registry as dr, entity_registry as er
+import homeassistant.helpers.config_validation as cv
+
+from .const import (
+    ADD_ENTITIES_CALLBACKS,
+    CONF_DOMAIN_DATA,
+    CONF_HARDWARE_SERIAL,
+    CONF_HARDWARE_TYPE,
+    CONF_SOFTWARE_SERIAL,
+    CONNECTION,
+    DOMAIN,
+)
+from .helpers import (
+    DeviceConnectionType,
+    async_update_device_config,
+    generate_unique_id,
+    get_device_config,
+    get_device_connection,
+    get_resource,
+    purge_device_registry,
+    purge_entity_registry,
+    register_lcn_address_devices,
+)
+from .schemas import (
+    ADDRESS_SCHEMA,
+    DOMAIN_DATA_BINARY_SENSOR,
+    DOMAIN_DATA_CLIMATE,
+    DOMAIN_DATA_COVER,
+    DOMAIN_DATA_LIGHT,
+    DOMAIN_DATA_SCENE,
+    DOMAIN_DATA_SENSOR,
+    DOMAIN_DATA_SWITCH,
+)
+
+URL_BASE: Final = "/lcn_static"
+
+
+async def register_panel(hass: HomeAssistant) -> None:
+    """Register the LCN Panel and Websocket API."""
+    websocket_api.async_register_command(hass, websocket_get_hosts)
+    websocket_api.async_register_command(hass, websocket_get_device_configs)
+    websocket_api.async_register_command(hass, websocket_get_entity_configs)
+    websocket_api.async_register_command(hass, websocket_scan_devices)
+    websocket_api.async_register_command(hass, websocket_add_device)
+    websocket_api.async_register_command(hass, websocket_delete_device)
+    websocket_api.async_register_command(hass, websocket_add_entity)
+    websocket_api.async_register_command(hass, websocket_delete_entity)
+
+    if DOMAIN not in hass.data.get("frontend_panels", {}):
+        hass.http.register_static_path(
+            URL_BASE,
+            path=lcn_panel.locate_dir(),
+            cache_headers=lcn_panel.is_prod_build,
+        )
+        await panel_custom.async_register_panel(
+            hass=hass,
+            frontend_url_path=DOMAIN,
+            webcomponent_name=lcn_panel.webcomponent_name,
+            config_panel_domain=DOMAIN,
+            module_url=f"{URL_BASE}/{lcn_panel.entrypoint_js}",
+            embed_iframe=True,
+            require_admin=True,
+        )
+
+
+@websocket_api.require_admin
+@websocket_api.websocket_command({vol.Required("type"): "lcn/hosts"})
+@websocket_api.async_response
+async def websocket_get_hosts(
+    hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
+) -> None:
+    """Get LCN hosts."""
```

Comment:
"get hosts" does not seem to match the implementation, the command returns a list of config entries?

Reply:
No, it sends a list of dictionary with names and ids of the available hosts. I guess it's correct as it is.

Reply:
I just realized that the config_entry_id is passed from the integration-panel to the custom-panel via url-parameter. This means that I do not need the host-selector box in the first screenshot. I therefore removed the whole `websocket_get_hosts` method and handle the url-parameter in the lcn-frontend code (https://github.com/alengwenus/lcn-frontend/compare/0.1.3...0.1.4)